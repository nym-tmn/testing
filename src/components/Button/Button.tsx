import type { MouseEventHandler, ReactNode } from "react"
import styles from './Button.module.css'

interface ButtonProps {
	children: ReactNode;
	disabled?: boolean;
	onClick?: MouseEventHandler<HTMLButtonElement>;
	type?: "submit" | "reset" | "button";
}

export const Button: React.FC<ButtonProps> = ({ children, onClick, disabled, type = 'button' }) => {

	return (
		<button
			className={styles.button}
			data-testid={'click-btn'}
			onClick={onClick}
			type={type}
			disabled={type === 'submit' ? disabled : false}
		>
			{disabled ? 'Processing...' : children}
		</button>
	)
}