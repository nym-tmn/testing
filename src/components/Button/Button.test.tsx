import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Button } from './Button';
import userEvent from '@testing-library/user-event';

describe('Component Button', () => {
	const handleClick = vi.fn();
	it('show text button', () => {
		render(<Button onClick={handleClick}>Press me</Button>);
		expect(screen.getByTestId('click-btn')).toBeInTheDocument();
	});

	it('calls onClick when clicked', async() => {
		const handleClick = vi.fn();
		render(<Button onClick={handleClick}>Press me</Button>);
		await userEvent.click(screen.getByTestId('click-btn'));
		expect(handleClick).toHaveBeenCalledTimes(1);
	})

	it('should be disabled if prop disabled is passed', async() => {
		const handleClick = vi.fn();
		render(<Button onClick={handleClick} disabled>Press me</Button>);
		expect(screen.getByTestId('click-btn')).toBeDisabled();
	})

	it('click should not work if the prop is disabled', async() => {
		const handleClick = vi.fn();
		render(<Button onClick={handleClick} disabled>Press me</Button>);
		await userEvent.click(screen.getByTestId('click-btn'));
		expect(handleClick).not.toHaveBeenCalled();
	})

	it('snapshot: Button with onClick and disabled', async() => {
		const handleClick = vi.fn();
		render(<Button onClick={handleClick} disabled>Press me</Button>);
		expect(screen.getByTestId('click-btn')).toMatchSnapshot();
	})
});