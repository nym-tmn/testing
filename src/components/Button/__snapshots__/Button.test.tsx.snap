// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component Button > snapshot 1`] = `
<button
  data-testid="click-btn"
  disabled=""
  type="button"
>
  Processing...
</button>
`;
